pycbc/waveform/generator.py

```diff
@@ -1225,30 +1229,57 @@ def select_waveform_generator(approximant):
+                'FdQNMfromFinalMassSpin': FDomainMassSpinRingdownGenerator,
+                'FdQNMfromFreqTau': FDomainFreqTauRingdownGenerator,
+            }
+        },
+        'td': {
+            'cbc': TDomainCBCGenerator,
+            'ringdown': {
+                'TdQNMfromFinalMassSpin': TDomainMassSpinRingdownGenerator,
+                'TdQNMfromFreqTau': TDomainFreqTauRingdownGenerator,
+            },
+            'supernovae': {
+                'CoreCollapseBounce': TDomainSupernovaeGenerator
+            }
+        }
+    }
+
+    gen = None
+    if domain not in {None, 'td', 'fd'}:
+        raise ValueError(f"Invalid domain '{domain}'. Must be one of: None, 'td', or 'fd'.")
+
+    # Check if the approximant exists in any domain
     if approximant in waveform.fd_approximants():
-        return FDomainCBCGenerator
-    # check if time-domain CBC waveform
+        if domain == 'fd' or domain is None:
+            gen = generator_map['fd']['cbc']
+        elif domain == 'td' and approximant in waveform.td_approximants():
+            gen = generator_map['td']['cbc']
+        elif domain == 'td':
+            raise ValueError(f"Requested domain is 'td', but {approximant} is a frequency-domain model.")
     elif approximant in waveform.td_approximants():
-        return TDomainCBCGenerator
-    # check if frequency-domain ringdown waveform
+        if domain == 'td' or domain is None:
+            gen = generator_map['td']['cbc']
+        elif domain == 'fd' and approximant in waveform.fd_approximants():
+            gen = generator_map['fd']['cbc']
+        elif domain == 'fd':
+            raise ValueError(f"Requested domain is 'fd', but {approximant} is a time-domain model.")
     elif approximant in ringdown.ringdown_fd_approximants:
-        if approximant == 'FdQNMfromFinalMassSpin':
-            return FDomainMassSpinRingdownGenerator
-        elif approximant == 'FdQNMfromFreqTau':
-            return FDomainFreqTauRingdownGenerator
+        gen = generator_map['fd']['ringdown'].get(approximant, None)
     elif approximant in ringdown.ringdown_td_approximants:
-        if approximant == 'TdQNMfromFinalMassSpin':
-            return TDomainMassSpinRingdownGenerator
-        elif approximant == 'TdQNMfromFreqTau':
-            return TDomainFreqTauRingdownGenerator
-    # check if supernovae waveform:
+        gen = generator_map['td']['ringdown'].get(approximant, None)
     elif approximant in supernovae.supernovae_td_approximants:
-        if approximant == 'CoreCollapseBounce':
-            return TDomainSupernovaeGenerator
-    # otherwise waveform approximant is not supported
-    else:
-        raise ValueError("%s is not a valid approximant." % approximant)
+        gen = generator_map['td']['supernovae'].get(approximant, None)
+
+    if gen is None:
+        raise ValueError(f"{approximant} is not a valid approximant.")
+
+    return gen
 
 
 def select_waveform_modes_generator(approximant):
```

Comment:
Could you add the domain argument to this function too, for consistency? Use the same logic you use above. You can use the `get_td_generator` and `get_fd_generator` functions I suggested writing above, just add a `modes=False` keyword argument to them so that they get the modes generators for CBC waveforms if `modes` is set to True. Then in this function you would do the same logic, just with `modes=True`.

Reply:
I can do that as well